components/hab/src/cli.rs

```diff
@@ -30,16 +31,192 @@ use habitat_core::{crypto::{keys::PairType,
+    /// /home/core-redis-3.0.7-21120102031201-x86_64-linux.hart)
+    #[structopt(name = "PKG_IDENT_OR_ARTIFACT")]
+    pkg_ident_or_artifact: Option<String>,
+    // TODO (DM): This flag can eventually be removed.
+    // See https://github.com/habitat-sh/habitat/issues/7339
+    #[structopt(name = "APPLICATION", long = "application", hidden = true)]
+    application: Vec<String>,
+    // TODO (DM): This flag can eventually be removed.
+    // See https://github.com/habitat-sh/habitat/issues/7339
+    #[structopt(name = "ENVIRONMENT", long = "environment", hidden = true)]
+    environment: Vec<String>,
+    /// The service group; shared config and topology [default: default]
+    // TODO (DM): This should set a default value
+    #[structopt(name = "GROUP", long = "group")]
+    group: String,
+    /// Service topology; [default: none]
+    // TODO (DM): I dont think saying the default is none makes sense here
+    #[structopt(name = "TOPOLOGY",
+                long = "topology",
+                short = "t",
+                possible_values = &["standalone", "leader"])]
+    topology: Option<habitat_sup_protocol::types::Topology>,
+    /// The update strategy; [default: none] [values: none, at-once, rolling]
+    // TODO (DM): this should use possible_values = &["none", "at-once", "rolling"]
+    #[structopt(name = "STRATEGY", long = "strategy", short = "s")]
+    strategy: Option<habitat_sup_protocol::types::UpdateStrategy>,
+    /// One or more service groups to bind to a configuration
+    #[structopt(name = "BIND", long = "bind")]
+    bind: Vec<String>,
+    /// Governs how the presence or absence of binds affects service startup. `strict` blocks
+    /// startup until all binds are present. [default: strict] [values: relaxed, strict]
+    // TODO (DM): values and default
+    #[structopt(name = "BINDING_MODE", long = "binding-mode")]
+    binding_mode: Option<habitat_sup_protocol::types::BindingMode>,
+    /// Verbose output; shows file and line/column numbers
+    #[structopt(name = "VERBOSE", short = "v")]
+    verbose: bool,
+    /// Turn ANSI color off
+    #[structopt(name = "NO_COLOR", long = "no-color")]
+    no_color: bool,
+    /// Use structured JSON logging for the Supervisor. Implies NO_COLOR
+    #[structopt(name = "JSON", long = "json-logging")]
+    json_logging: bool,
+    /// The interval (seconds) on which to run health checks [default: 30]
+    // TODO (DM): Should use default_value = "30"
+    #[structopt(name = "HEALTH_CHECK_INTERVAL",
+                long = "health-check-interval",
+                short = "i")]
+    health_check_interval: HealthCheckInterval,
+    /// The IPv4 address to use as the `sys.ip` template variable. If this argument is not set, the
+    /// supervisor tries to dynamically determine an IP address. If that fails, the supervisor
+    /// defaults to using `127.0.0.1`[.]
+    #[structopt(name = "SYS_IP_ADDRESS", long = "sys-ip-address")]
+    sys_ip_address: Option<Ipv4Addr>,
+}
```

Comment:
I think `structopt`'s [flattening](https://docs.rs/structopt/0.3.8/structopt/index.html#flattening) feature was intended to exactly address this issue. Instead of the unit of composition being functions, it would now be struct definitions.

What is your opinion of `structopt`? I am kinda of proposing we switch all cli arg definitions to use it. Does that seem sane? Personally, I am a huge fan of it, but would like to make sure others are on board :smiley:.

Reply:
Might get weird if we have to have a bunch of one-field structs (at the extreme end of things), but there are probably worse things.

Our CLI definitions have historically been a pain point, so I'm absolutely in favor of anything that helps bring sanity to it. It seems like this could do that, in addition to unlocking some nice functionality elsewhere in the codebase. 

I'd love to see where this leads
![](https://media.giphy.com/media/Hitk8jG5sCHe0/giphy.gif)

Reply:
I think the direction you're headed @davidMcneil with `structopt` is spot on, even more-so with the advent of the flattening feature @davidMcneil!

![eyebrow](https://user-images.githubusercontent.com/1991696/73114653-7f770f80-3eea-11ea-98ea-6dde9ac49dd6.gif)